File: test/py-stub-either-overlay/BaboonTests/WiringTests.py
import unittest
import json
from io import BytesIO

from BaboonDefinitions.Generated.baboon_service_wiring import (
    BaboonMethodId, BaboonLeft, BaboonRight,
    NoMatchingMethod, CallFailed, DecoderFailed,
    JsonMuxer, UebaMuxer, BaboonWiringException, NoMatchingService, DuplicateService,
)
from BaboonDefinitions.Generated.baboon_codecs import BaboonCodecContext
from BaboonDefinitions.Generated.baboon_runtime_shared import LEDataInputStream, LEDataOutputStream
from BaboonDefinitions.Generated.testpkg.pkg0.BaboonServiceRt import BaboonServiceRtDefault
from BaboonDefinitions.Generated.testpkg.pkg0.I1 import I1
from BaboonDefinitions.Generated.testpkg.pkg0.I2 import I2
from BaboonDefinitions.Generated.testpkg.pkg0.I1_Wiring import invoke_json_I1, invoke_ueba_I1, I1_JsonService, I1_UebaService
from BaboonDefinitions.Generated.testpkg.pkg0.I2_Wiring import invoke_json_I2, invoke_ueba_I2, I2_JsonService, I2_UebaService
from BaboonDefinitions.Generated.testpkg.pkg0.i1.testcall.In import In as I1_In, In_UEBACodec as I1_In_UEBACodec
from BaboonDefinitions.Generated.testpkg.pkg0.i1.testcall.Out import Out as I1_Out, Out_UEBACodec as I1_Out_UEBACodec
from BaboonDefinitions.Generated.testpkg.pkg0.i1.testcall.Err import Err as I1_Err
from BaboonDefinitions.Generated.testpkg.pkg0.T7_Empty import T7_Empty, T7_Empty_UEBACodec
from BaboonDefinitions.Generated.testpkg.pkg0.i2.noerrcall.In import In as I2_In, In_UEBACodec as I2_In_UEBACodec
from BaboonDefinitions.Generated.testpkg.pkg0.i2.noerrcall.Out import Out as I2_Out, Out_UEBACodec as I2_Out_UEBACodec


class MockI1(I1):
    def testCall(self, arg):
        return BaboonRight(I1_Out(i00=42))

    def testCall2(self, arg):
        return BaboonRight(T7_Empty())


class FailingI1(I1):
    def testCall(self, arg):
        return BaboonLeft(I1_Err(msg="domain error"))

    def testCall2(self, arg):
        return BaboonLeft(T7_Empty())


class ThrowingI1(I1):
    def testCall(self, arg):
        raise RuntimeError("service error")

    def testCall2(self, arg):
        raise RuntimeError("service error")


class MockI2(I2):
    def noErrCall(self, arg):
        return I2_Out(result="result_" + str(arg.value))


class EitherWiringTests(unittest.TestCase):

    ctx = BaboonCodecContext.default()
    rt = BaboonServiceRtDefault()

    # ==================== I1 JSON Tests ====================

    def test_i1_json_testCall_success(self):
        impl = MockI1()
        method = BaboonMethodId("I1", "testCall")
        result = invoke_json_I1(method, "{}", impl, self.rt, self.ctx)
        self.assertIsInstance(result, BaboonRight)
        decoded = json.loads(result.value)
        self.assertEqual(decoded["i00"], 42)

    def test_i1_json_testCall2_success(self):
        impl = MockI1()
        method = BaboonMethodId("I1", "testCall2")
        result = invoke_json_I1(method, "{}", impl, self.rt, self.ctx)
        self.assertIsInstance(result, BaboonRight)

    def test_i1_json_domain_error(self):
        impl = FailingI1()
        method = BaboonMethodId("I1", "testCall")
        result = invoke_json_I1(method, "{}", impl, self.rt, self.ctx)
        self.assertIsInstance(result, BaboonLeft)
        self.assertIsInstance(result.value, CallFailed)

    def test_i1_json_no_matching_method(self):
        impl = MockI1()
        method = BaboonMethodId("I1", "nonexistent")
        result = invoke_json_I1(method, "{}", impl, self.rt, self.ctx)
        self.assertIsInstance(result, BaboonLeft)
        self.assertIsInstance(result.value, NoMatchingMethod)

    def test_i1_json_decoder_failed(self):
        impl = MockI1()
        method = BaboonMethodId("I1", "testCall")
        result = invoke_json_I1(method, "not valid json!!", impl, self.rt, self.ctx)
        self.assertIsInstance(result, BaboonLeft)
        self.assertIsInstance(result.value, DecoderFailed)

    def test_i1_json_service_throws(self):
        impl = ThrowingI1()
        method = BaboonMethodId("I1", "testCall")
        result = invoke_json_I1(method, "{}", impl, self.rt, self.ctx)
        self.assertIsInstance(result, BaboonLeft)
        self.assertIsInstance(result.value, CallFailed)

    # ==================== I1 UEBA Tests ====================

    def test_i1_ueba_testCall_success(self):
        impl = MockI1()
        method = BaboonMethodId("I1", "testCall")
        buf = BytesIO()
        writer = LEDataOutputStream(buf)
        I1_In_UEBACodec.instance().encode(self.ctx, writer, I1_In())
        input_bytes = buf.getvalue()
        result = invoke_ueba_I1(method, input_bytes, impl, self.rt, self.ctx)
        self.assertIsInstance(result, BaboonRight)
        reader = LEDataInputStream(BytesIO(result.value))
        decoded = I1_Out_UEBACodec.instance().decode(self.ctx, reader)
        self.assertEqual(decoded.i00, 42)

    def test_i1_ueba_no_matching_method(self):
        impl = MockI1()
        method = BaboonMethodId("I1", "nonexistent")
        result = invoke_ueba_I1(method, b'', impl, self.rt, self.ctx)
        self.assertIsInstance(result, BaboonLeft)
        self.assertIsInstance(result.value, NoMatchingMethod)

    def test_i1_ueba_service_throws(self):
        impl = ThrowingI1()
        method = BaboonMethodId("I1", "testCall")
        buf = BytesIO()
        writer = LEDataOutputStream(buf)
        I1_In_UEBACodec.instance().encode(self.ctx, writer, I1_In())
        input_bytes = buf.getvalue()
        result = invoke_ueba_I1(method, input_bytes, impl, self.rt, self.ctx)
        self.assertIsInstance(result, BaboonLeft)
        self.assertIsInstance(result.value, CallFailed)

    # ==================== I2 JSON Tests ====================

    def test_i2_json_success(self):
        impl = MockI2()
        method = BaboonMethodId("I2", "noErrCall")
        result = invoke_json_I2(method, '{"value": 123}', impl, self.rt, self.ctx)
        self.assertIsInstance(result, BaboonRight)
        decoded = json.loads(result.value)
        self.assertEqual(decoded["result"], "result_123")

    # ==================== I2 UEBA Tests ====================

    def test_i2_ueba_success(self):
        impl = MockI2()
        method = BaboonMethodId("I2", "noErrCall")
        buf = BytesIO()
        writer = LEDataOutputStream(buf)
        I2_In_UEBACodec.instance().encode(self.ctx, writer, I2_In(value=456))
        input_bytes = buf.getvalue()
        result = invoke_ueba_I2(method, input_bytes, impl, self.rt, self.ctx)
        self.assertIsInstance(result, BaboonRight)
        reader = LEDataInputStream(BytesIO(result.value))
        decoded = I2_Out_UEBACodec.instance().decode(self.ctx, reader)
        self.assertEqual(decoded.result, "result_456")


# ==================== Cross-domain Muxer ====================
# A single muxer composes the I1 (errors mode) and I2 (no-err mode)
# services and routes each call by method.service_name.

class MuxerWiringTests(unittest.TestCase):

    ctx = BaboonCodecContext.default()
    rt = BaboonServiceRtDefault()

    def _json_muxer(self):
        return JsonMuxer(
            I1_JsonService(MockI1(), self.rt),
            I2_JsonService(MockI2(), self.rt),
        )

    def _ueba_muxer(self):
        return UebaMuxer(
            I1_UebaService(MockI1(), self.rt),
            I2_UebaService(MockI2(), self.rt),
        )

    def test_json_muxer_routes_to_i1(self):
        method = BaboonMethodId("I1", "testCall")
        result = self._json_muxer().invoke(method, "{}", self.ctx)
        self.assertIsInstance(result, BaboonRight)
        self.assertEqual(json.loads(result.value)["i00"], 42)

    def test_json_muxer_routes_to_i2(self):
        method = BaboonMethodId("I2", "noErrCall")
        result = self._json_muxer().invoke(method, '{"value": 123}', self.ctx)
        self.assertIsInstance(result, BaboonRight)
        self.assertEqual(json.loads(result.value)["result"], "result_123")

    def test_json_muxer_no_matching_service(self):
        method = BaboonMethodId("Nonexistent", "x")
        with self.assertRaises(BaboonWiringException) as cm:
            self._json_muxer().invoke(method, "{}", self.ctx)
        self.assertIsInstance(cm.exception.error, NoMatchingService)

    def test_json_muxer_duplicate_service(self):
        with self.assertRaises(BaboonWiringException) as cm:
            JsonMuxer(
                I1_JsonService(MockI1(), self.rt),
                I1_JsonService(MockI1(), self.rt),
            )
        self.assertIsInstance(cm.exception.error, DuplicateService)

    def test_ueba_muxer_routes_to_i1(self):
        method = BaboonMethodId("I1", "testCall")
        buf = BytesIO()
        I1_In_UEBACodec.instance().encode(self.ctx, LEDataOutputStream(buf), I1_In())
        result = self._ueba_muxer().invoke(method, buf.getvalue(), self.ctx)
        self.assertIsInstance(result, BaboonRight)
        decoded = I1_Out_UEBACodec.instance().decode(self.ctx, LEDataInputStream(BytesIO(result.value)))
        self.assertEqual(decoded.i00, 42)

    def test_ueba_muxer_routes_to_i2(self):
        method = BaboonMethodId("I2", "noErrCall")
        buf = BytesIO()
        I2_In_UEBACodec.instance().encode(self.ctx, LEDataOutputStream(buf), I2_In(value=456))
        result = self._ueba_muxer().invoke(method, buf.getvalue(), self.ctx)
        self.assertIsInstance(result, BaboonRight)
        decoded = I2_Out_UEBACodec.instance().decode(self.ctx, LEDataInputStream(BytesIO(result.value)))
        self.assertEqual(decoded.result, "result_456")

    def test_ueba_muxer_no_matching_service(self):
        method = BaboonMethodId("Nonexistent", "x")
        with self.assertRaises(BaboonWiringException) as cm:
            self._ueba_muxer().invoke(method, b"", self.ctx)
        self.assertIsInstance(cm.exception.error, NoMatchingService)


if __name__ == "__main__":
    unittest.main()

File: test/rs-stub-either-overlay/tests/wiring_tests.rs
use baboon_rs_stub::baboon_runtime::{BaboonCodecContext, BaboonBinEncode, BaboonBinDecode};
use baboon_rs_stub::baboon_service_wiring::{BaboonMethodId, BaboonWiringError, JsonMuxer, UebaMuxer};
use baboon_rs_stub::testpkg::pkg0::baboon_service_rt::BaboonServiceRtDefault;
use baboon_rs_stub::testpkg::pkg0::i1::I1;
use baboon_rs_stub::testpkg::pkg0::i1::testcall::input::In as I1_testCall_in;
use baboon_rs_stub::testpkg::pkg0::i1::testcall::out::Out as I1_testCall_out;
use baboon_rs_stub::testpkg::pkg0::i1::testcall::err::Err as I1_testCall_err;
use baboon_rs_stub::testpkg::pkg0::i2::I2;
use baboon_rs_stub::testpkg::pkg0::i2::noerrcall::input::In as I2_noErrCall_in;
use baboon_rs_stub::testpkg::pkg0::i2::noerrcall::out::Out as I2_noErrCall_out;
use baboon_rs_stub::testpkg::pkg0::t7_empty::T7_Empty;
use baboon_rs_stub::testpkg::pkg0::i1_wiring::{invoke_json_i1, invoke_ueba_i1, I1JsonService, I1UebaService};
use baboon_rs_stub::testpkg::pkg0::i2_wiring::{invoke_json_i2, invoke_ueba_i2, I2JsonService, I2UebaService};

struct MockI1;
impl I1 for MockI1 {
    fn test_call(&self, _arg: I1_testCall_in) -> Result<I1_testCall_out, I1_testCall_err> {
        Ok(I1_testCall_out { i00: 42 })
    }
    fn test_call2(&self, _arg: T7_Empty) -> Result<T7_Empty, T7_Empty> {
        Ok(T7_Empty {})
    }
}

struct FailingI1;
impl I1 for FailingI1 {
    fn test_call(&self, _arg: I1_testCall_in) -> Result<I1_testCall_out, I1_testCall_err> {
        Err(I1_testCall_err { msg: "domain error".to_string() })
    }
    fn test_call2(&self, _arg: T7_Empty) -> Result<T7_Empty, T7_Empty> {
        Err(T7_Empty {})
    }
}

struct MockI2;
impl I2 for MockI2 {
    fn no_err_call(&self, arg: I2_noErrCall_in) -> I2_noErrCall_out {
        I2_noErrCall_out { result: format!("result_{}", arg.value) }
    }
}

#[test]
fn i1_json_test_call_success() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "testCall".to_string() };
    let input_json = serde_json::to_string(&I1_testCall_in {}).unwrap();
    let rt = BaboonServiceRtDefault;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_json_i1(&method, &input_json, &MockI1, &rt, &ctx);
    assert!(result.is_ok());
    let decoded: I1_testCall_out = serde_json::from_str(&result.unwrap()).unwrap();
    assert_eq!(decoded.i00, 42);
}

#[test]
fn i1_json_test_call2_success() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "testCall2".to_string() };
    let input_json = serde_json::to_string(&T7_Empty {}).unwrap();
    let rt = BaboonServiceRtDefault;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_json_i1(&method, &input_json, &MockI1, &rt, &ctx);
    assert!(result.is_ok());
}

#[test]
fn i1_json_domain_error() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "testCall".to_string() };
    let input_json = serde_json::to_string(&I1_testCall_in {}).unwrap();
    let rt = BaboonServiceRtDefault;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_json_i1(&method, &input_json, &FailingI1, &rt, &ctx);
    assert!(result.is_err());
}

#[test]
fn i1_json_unknown_method() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "nonexistent".to_string() };
    let rt = BaboonServiceRtDefault;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_json_i1(&method, "{}", &MockI1, &rt, &ctx);
    assert!(result.is_err());
}

#[test]
fn i1_json_bad_input() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "testCall".to_string() };
    let rt = BaboonServiceRtDefault;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_json_i1(&method, "not valid json!!", &MockI1, &rt, &ctx);
    assert!(result.is_err());
}

#[test]
fn i1_ueba_test_call_success() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "testCall".to_string() };
    let ctx = BaboonCodecContext::Default;
    let rt = BaboonServiceRtDefault;
    let mut writer = Vec::new();
    BaboonBinEncode::encode_ueba(&I1_testCall_in {}, &ctx, &mut writer).unwrap();
    let result = invoke_ueba_i1(&method, &writer, &MockI1, &rt, &ctx);
    assert!(result.is_ok());
    let mut cursor = std::io::Cursor::new(result.unwrap());
    let decoded: I1_testCall_out = BaboonBinDecode::decode_ueba(&ctx, &mut cursor).unwrap();
    assert_eq!(decoded.i00, 42);
}

#[test]
fn i1_ueba_unknown_method() {
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "nonexistent".to_string() };
    let rt = BaboonServiceRtDefault;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_ueba_i1(&method, &[], &MockI1, &rt, &ctx);
    assert!(result.is_err());
}

#[test]
fn i2_json_no_err_call_success() {
    let method = BaboonMethodId { service_name: "I2".to_string(), method_name: "noErrCall".to_string() };
    let input_json = serde_json::to_string(&I2_noErrCall_in { value: 123 }).unwrap();
    let rt = BaboonServiceRtDefault;
    let ctx = BaboonCodecContext::Default;
    let result = invoke_json_i2(&method, &input_json, &MockI2, &rt, &ctx);
    assert!(result.is_ok());
    let decoded: I2_noErrCall_out = serde_json::from_str(&result.unwrap()).unwrap();
    assert_eq!(decoded.result, "result_123");
}

#[test]
fn i2_ueba_no_err_call_success() {
    let method = BaboonMethodId { service_name: "I2".to_string(), method_name: "noErrCall".to_string() };
    let ctx = BaboonCodecContext::Default;
    let rt = BaboonServiceRtDefault;
    let mut writer = Vec::new();
    BaboonBinEncode::encode_ueba(&I2_noErrCall_in { value: 456 }, &ctx, &mut writer).unwrap();
    let result = invoke_ueba_i2(&method, &writer, &MockI2, &rt, &ctx);
    assert!(result.is_ok());
    let mut cursor = std::io::Cursor::new(result.unwrap());
    let decoded: I2_noErrCall_out = BaboonBinDecode::decode_ueba(&ctx, &mut cursor).unwrap();
    assert_eq!(decoded.result, "result_456");
}

// ==================== Cross-domain Muxer ====================
// A single muxer composes the I1 (errors mode) and I2 (no-err mode)
// services and routes each call by method.service_name. The outer Result
// is the muxer's routing result; the inner is the service's own result.

type JsonR = Result<String, BaboonWiringError>;
type UebaR = Result<Vec<u8>, BaboonWiringError>;

fn new_json_muxer() -> JsonMuxer<JsonR> {
    JsonMuxer::<JsonR>::new()
        .with(Box::new(I1JsonService::new(MockI1, BaboonServiceRtDefault))).unwrap()
        .with(Box::new(I2JsonService::new(MockI2, BaboonServiceRtDefault))).unwrap()
}

fn new_ueba_muxer() -> UebaMuxer<UebaR> {
    UebaMuxer::<UebaR>::new()
        .with(Box::new(I1UebaService::new(MockI1, BaboonServiceRtDefault))).unwrap()
        .with(Box::new(I2UebaService::new(MockI2, BaboonServiceRtDefault))).unwrap()
}

#[test]
fn json_muxer_routes_to_i1() {
    let ctx = BaboonCodecContext::Default;
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "testCall".to_string() };
    let input_json = serde_json::to_string(&I1_testCall_in {}).unwrap();
    let routed = new_json_muxer().invoke(&method, &input_json, &ctx).expect("routing");
    let decoded: I1_testCall_out = serde_json::from_str(&routed.unwrap()).unwrap();
    assert_eq!(decoded.i00, 42);
}

#[test]
fn json_muxer_routes_to_i2() {
    let ctx = BaboonCodecContext::Default;
    let method = BaboonMethodId { service_name: "I2".to_string(), method_name: "noErrCall".to_string() };
    let input_json = serde_json::to_string(&I2_noErrCall_in { value: 123 }).unwrap();
    let routed = new_json_muxer().invoke(&method, &input_json, &ctx).expect("routing");
    let decoded: I2_noErrCall_out = serde_json::from_str(&routed.unwrap()).unwrap();
    assert_eq!(decoded.result, "result_123");
}

#[test]
fn json_muxer_no_matching_service() {
    let ctx = BaboonCodecContext::Default;
    let method = BaboonMethodId { service_name: "Nonexistent".to_string(), method_name: "x".to_string() };
    let routed = new_json_muxer().invoke(&method, "{}", &ctx);
    assert!(matches!(routed, Err(BaboonWiringError::NoMatchingService(_))));
}

#[test]
fn json_muxer_duplicate_service() {
    let dup = JsonMuxer::<JsonR>::new()
        .with(Box::new(I1JsonService::new(MockI1, BaboonServiceRtDefault))).unwrap()
        .with(Box::new(I1JsonService::new(MockI1, BaboonServiceRtDefault)));
    assert!(matches!(dup, Err(BaboonWiringError::DuplicateService(_))));
}

#[test]
fn ueba_muxer_routes_to_i1() {
    let ctx = BaboonCodecContext::Default;
    let method = BaboonMethodId { service_name: "I1".to_string(), method_name: "testCall".to_string() };
    let mut writer = Vec::new();
    BaboonBinEncode::encode_ueba(&I1_testCall_in {}, &ctx, &mut writer).unwrap();
    let routed = new_ueba_muxer().invoke(&method, &writer, &ctx).expect("routing");
    let mut cursor = std::io::Cursor::new(routed.unwrap());
    let decoded: I1_testCall_out = BaboonBinDecode::decode_ueba(&ctx, &mut cursor).unwrap();
    assert_eq!(decoded.i00, 42);
}

#[test]
fn ueba_muxer_routes_to_i2() {
    let ctx = BaboonCodecContext::Default;
    let method = BaboonMethodId { service_name: "I2".to_string(), method_name: "noErrCall".to_string() };
    let mut writer = Vec::new();
    BaboonBinEncode::encode_ueba(&I2_noErrCall_in { value: 456 }, &ctx, &mut writer).unwrap();
    let routed = new_ueba_muxer().invoke(&method, &writer, &ctx).expect("routing");
    let mut cursor = std::io::Cursor::new(routed.unwrap());
    let decoded: I2_noErrCall_out = BaboonBinDecode::decode_ueba(&ctx, &mut cursor).unwrap();
    assert_eq!(decoded.result, "result_456");
}

#[test]
fn ueba_muxer_no_matching_service() {
    let ctx = BaboonCodecContext::Default;
    let method = BaboonMethodId { service_name: "Nonexistent".to_string(), method_name: "x".to_string() };
    let routed = new_ueba_muxer().invoke(&method, &[], &ctx);
    assert!(matches!(routed, Err(BaboonWiringError::NoMatchingService(_))));
}
